### DISCUSSION TITLE
Intellisense documentation not working when sourcing/running other notebooks

### DISCUSSION THREAD

**Discussion by mwh3780** [microsoft/pylance-release#2037] [Q&A] [answered]
I recently inherited a process where code is spread out and stored in various Jupyter notebooks and they're linked together by calling each other through `%run`ning / sourcing each other.

In trying to use my normal Jupyter setup in VS Code with Pylance as the language server I've noticed that Intellisense doesn't completely work. For objects and functions defined in notebooks that are sourced as well as any modules imported in these notebook, autocompletion does work when typing the names of these objects/functions/modules, but the hover over documentation doesn't. When I hover over any of these objects, whether they are custom objects, functions, methods or modules that are defined in notebooks that are sourced the hover over documentation simply shows `Object name: Any`. And just for reference, JupyterLab and classic Jupyter Notebook don't have this issue.

I tried looking for help or troubleshooting guides for this particular issue, but didn't find anything that helped.

Since reproducing the behavior involves different files, below is some minimal code that shows the behavior:

```
### Code stored in file named random_code.ipynb
import pandas as pd
import numpy as np

df_1 = pd.DataFrame(np.random.randint(0, 10, size = (3, 3)), columns = ["A", "B", "C"])

def fun1(x, y = 1):
    '''
    Some function

    Arguments:
    ...

    Examples:
    ...
    '''
    return(x + y)
```

```
%run "/Users/mwh3780/Box/DataExplorer/Test/hammer_sandbox/Sub folder/random_code.ipynb"

# Hover over documentation does not work for any of the following :
df_1.sort_values()
fun1(1, 2)
pd.DataFrame()
```

Sees the object in the environment
<img width="579" alt="image" src="https://user-images.githubusercontent.com/23129139/140347672-0b82f0bc-869c-4259-9ab7-d73407c50495.png">

But documentation doesn't work
<img width="404" alt="image" src="https://user-images.githubusercontent.com/23129139/140347043-d273e060-f51d-40b4-bba6-65a1d87e5cdc.png">

<img width="191" alt="image" src="https://user-images.githubusercontent.com/23129139/140356667-ab151ca4-c680-4e19-b9a8-2bb9d77c8c81.png">

<img width="185" alt="image" src="https://user-images.githubusercontent.com/23129139/140356778-95e6776c-8662-46aa-a8d7-1391ad0b7f83.png">

Any ideas?

I'm using:
OS: macOS Big Sur 11.4
VS Code: Version: 1.61.2 (Universal)
Jupyter Extension Version: v2021.9.1101343141
Pylance Extension Version: v2021.11.0
Python Extension Version: v2021.10.1365161279

**Comment by jakebailey** [comment, marked as answer]
Pylance is a static analyzer, and doesn't consume any sort of magic notebook state that these sorts of multi-notebook setups use. We just don't know anything but that one notebook, top to bottom.

Completion works because the Jupyter extension augments the completion list with at-runtime data. It can't do that for anything else, though.

We're currently working on improving Jupyter support, this would be worth an issue.

**Reply by mwh3780** [reply]
@jakebailey Ok, thank you. I'll create an issue for this then.
